src/main/java/Problem1/ValidParentheses.java

```diff
@@ -4,8 +4,43 @@
 
     // Do not change signature (function name, parameters, return type)
     public static boolean isValid(String str) {
-        // homework
-        // !!! must use ArrayStack or LinkedListStack from problem 1
-        return false; // place holder
+        LinkedListStack<Character> lefties = new LinkedListStack<Character>();
+
+        lefties.push('*'); //To avoid a NullPointerException, I have added a star
+
+        if (str == null) { //Speaking of null, the input String must be checked
+            return true;
+        }
+
+        for (int i = 0; i < str.length(); i++) { //Go through the entire input String
+
+            if (str.charAt(i) == ')') { //Check for matching round parenthesis on top of stack
+                if (lefties.peek() == '(') {
+                    lefties.pop();
+                } else {
+                    return false; //String must be invalid
+                }
+            } else if (str.charAt(i) == '}') { //Check for matching curly parenthesis on top of stack
+                if (lefties.peek() == '{') {
+                    lefties.pop();
+                } else {
+                    return false; //String must be invalid
+                }
+            } else if (str.charAt(i) == ']') { //Check for matching rectangular parenthesis on top of stack
+                if (lefties.peek() == '[') {
```

Comment:
the condition check here is quite complicated. this makes it hard to read and verify logic. try and see if this can refactored into less level of condition check.

Reply:
here's my suggestion:

```java
here's what I have for this part. just for reference and comparision.

```java 
   public static boolean isValid(String str) {
        if (str == null || str.isEmpty() || str.trim().length() == 0) {
            return true;
        }

        Stack<Character> stack = new LinkedListStack<>();

        for (int i = 0; i < str.length(); i++) {
            char c = str.charAt(i);
            Character left = stack.peek();
            if (left == null || !characterMatch(left, c)) {
                stack.push(c);
                continue;
            }
            stack.pop();
        }
        return stack.size() == 0;
    }

    private static boolean characterMatch(char left, char right) {
        switch (left) {
            case '(':
                return right == ')';
            case '{':
                return right == '}';
            case '[':
                return right == ']';
        }
        return false;
    }
```